[2기-P 김민성] JPA 게시판 미션 코드 리뷰 부탁드립니다.

epicblues opened the pull request:
# 📌 과제 설명 

 * JPA가 익숙치 않아서 학습 테스트 코드를 작성하면서 구현해 나갔습니다. 
 * 핵심 모듈(컨트롤러, 서비스)에 대한 테스트를 최대한 작성하려고 시도해봤습니다.
 * 예외 상황에 대한 테스트 코드 작성이 어려웠습니다.  😥
 * REST-DOCS 는 각 endpoint에 대한 성공 / 실패 case를 나누어 작성했습니다. 
     -> 사용자가 쉽게 접근할 수 있게 [정적 리소스 경로에 문서가 생성](https://github.com/epicblues/springboot-board-jpa/blob/main/src/main/resources/static/docs/post.html) 되도록 설정을 했습니다.

# **👩‍💻 요구 사항**

##  **✅SpringDataJPA 를 설정한다.** 

-  datasource : h2

## **✅엔티티를 구성한다**

  회원(User) 

     id (PK) (auto increment)
     name
     age
     hobby
     **created_at**
     **created_by**
 게시글(Post) 

     id (PK) (auto increment)
     title
     content
     **created_at**
     **created_by**
  
 회원과 게시글에 대한 연관관계를 설정한다. 

     회원과 게시글은 1:N 관계이다.
    
 게시글 Repository를 구현한다. (PostRepository) 

## **✅API를 구현한다.**

 게시글 조회

     페이징 조회 (GET "/posts")
     단건 조회 (GET "/posts/{id}")

 게시글 작성 (POST "/posts")

 게시글 수정 (POST "/posts/{id}")

## **✅REST-DOCS를 이용해서 문서화한다.**

#  PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->


#### 1. 멘토님은 메서드에서 예외가 발생하는 상황을 테스트 하실 때 미리 예외 클래스를 설계하시는지 궁금합니다! 


#### 2.. 예외 클래스를 작성하실 때 애플리케이션에서 최상위 예외 클래스를 먼저 작성하시는 지도 궁금합니다. 
* 저는 생각이 잘 나지 않아서 일단 모두 RuntimeException을 상속 받아서 작성했는데, 먼저 어느 정도 구조를 짜고 들어갔어야 했을까요?

#### 3. TODO 1 관련 : 시간을 직렬화할 때 format을 프로젝트 별로 따로 작성하시는지, 아니면 업계의 표준(?) 같이 자주 쓰는 형식이 있는지 궁금합니다!

#### 4. TODO 2 관련 : 테스트 전용 util 클래스에 대한 기능을 테스트하고 싶어서 해당 클래스의 내부 클래스에 테스트를 작성했는데, 멘토님은 따로 테스트 클래스 파일을 작성하시는지 궁금합니다!

#### 5. TODO 3 관련 : 저번 바우처 과제에서 Controller를 테스트 할 때 Service를 모의 객체를 만들어서 테스트 코드를 작성했으나, 이번 과제에서는 실제 작동하는 service를 이용해서 통합 테스트로 작성했습니다. 이유는 이렇습니다.

1. 이 테스트로 rest docs가 작성되기 되므로 이 테스트는 유닛 테스트에는 부적절한 것 같다.
2. 모의 객체에 대한 stubbing 관련 코드를 매번 작성해야 하는 불편하다.(사실 이게 제일 큽니다...☠️).

REST 문서 작성용(?) 통합 테스트, Controller 용 유닛 테스트를 따로 작성했어야 하는 걸까요?

epicblues committed:
chore : 프로젝트 초기화

epicblues committed:
chore : 테스트 관련 설정

TestJdbcConfiguration
* 테스트 시 JVM 라이프사이클을 따라가는 데이터베이스 컨테이너 설정
* 컨테이너 DataSource 빈 등록

application.yaml
* hibernate ddl-auto 설정

pom.xml
* p6spy 의존성

epicblues committed:
feat : Post,User 엔터티

* Post, User
* PostTest, UserTest
필드값 요구사항 충족

* EntityJpaTest
Jpa와 Entity 컨텍스트 등록 성공 여부 테스트

epicblues committed:
feat : 공통 필드 추상 클래스

CommonFieldEntity
* 생성 일자와, 작성자 필드를 가진 @MappedSuperclass
* 생성 일자는 Auditing을 통해 자동 생성

UserTest, PostTest
* Auditing 기능 테스트

epicblues committed:
feat : entity 연관관계

User, Post
UserTest, PostTest
* 양방향 연관관계 Many(Post) <-> One(User)
* 연관관계 설정 메서드
* 연관관계 제거 메서드

EntityJpaTest
* User, Post 연관 관계 JPA 연동 테스트

epicblues committed:
feat : PostRepository

PostRepository
PostRepositoryTest
* JpaRepository<Post,Long> 상속 인터페이스
* 학습 테스트

epicblues committed:
fix : Embedded DB 의존성 변경

TestContainer MySQL -> H2
테스트 시작 설정 오버헤드 감소

epicblues committed:
feat : PostService 조회, 생성

PostService, PostServiceTest
* 원하는 페이지만 조회하기
* 전체 조회
* 조회는 모두 생성 시간 기준 내림차순
* post 생성 기능
* 사용자 확인을 위해 UserRepository 의존성;

Post, Test
* equals, hashCode 추가
* 사용자 할당 시 createdBy 자동 등록

epicblues committed:
feat : Post 수정, 식별자 사용 조회

PostService, PostServiceTest
* UserRepository 의존성 추가
* Post의 식별자 id 사용 조회 기능 추가
* Post 수정 기능 추가

PostResponse
* PostService로 반환되는 Post의 Dto;
* 제목, 내용, 식별자 관리

epicblues committed:
feat : PostResponse 필드 추가

* PostResponse, PostServiceTest
createdAt, createdBy, authorId 필드 추가
변경에 따른 관련 테스트 수정

epicblues committed:
fix : id setter 변경

Post, User
* setId 제거

LongIdHolder 인터페이스
* long 타입 id 필드를 가진 인터페이스
* Post, User 가 구현

FieldModifier
* private long id를 가진 객체의 id 수정 기능

epicblues committed:
fix : Post 서비스 조회 반환 타입

* PostService, PostServiceTest
다중 Post 조회 메서드가 PostResponse 리스트를 반환하도록 변경

* FieldSetter#FieldSetterTest
FieldSetterTest가 제대로 작동하도록 내부 정적 클래스로 변경

epicblues committed:
feat : 게시글 페이징 조회

PostController, PostControllerTest
* GET /posts?page={pageNumber}&size={sizeNumber} 요청 처리
* 두 파라미터의 조건 만족시키지 못할 경우 400 Bad Request 응답
* 페이지 요청에 맞는 Post 목록 반환

PostPagePayload
* 페이지 요청용 Dto
* 요청 파라미터 검증

InvalidRequestArgumentException
* 요청 파라미터 검증 실패시 발생시키는 예외
* 현재 PostController에서 처리

FormatConfiguration
* 날짜 직렬화 설정

epicblues committed:
feat : 단일 게시글 조회

PostController, PostControllerTest
GET /posts/{postId}
* 단일 게시글 조회
* 게시글이 없을 경우 404 Not Found 응답

ResourceNotFoundException
* 단일 리소스가 존재하지 않을 때 Controller에서 던지는 예외

RestControllerExceptionHandler
* RestController에서 발생하는 예외 처리
 1. ResourceNotFoundException
 2. InvalidRequestArgumentException

epicblues committed:
test : 테스트 유틸 패키지 이름 변경

epicblues committed:
feat : 게시글 작성, 수정

PostController, PostControllerTest

POST /posts {userId:int, title:string, content:string}
* 게시글 작성
* 작성 성공 시 게시글 세부 정보 반환

POST /posts/{postId} {title:string, content:string}
* 게시글 수정
* 수정 성공 시 게시글 세부 정보 반환

PostRequest
* nullable한 userId 필드 추가(필드에 값이 들어올 때만 검증)

epicblues committed:
feat : REST docs 설정 및 작성

post.adoc -> post.html
* PostController 관련 API REST docs 작성 스크립트
* /src/resource/static/docs에 저장

pom.xml
* REST docs html 생성 관련 플러그인 설정

PostControllerTest
* REST docs 문서화 작성 스크립트 추가

EntityJpaTest
* 트랜잭션 초기화 tearDown ㅊ가

epicblues committed:
chore : PR용 TODO indexing

epicblues committed:
test : dto 클래스 유효성 검증

res-cogitans commented on .gitignore:
```diff
@@ -0,0 +1,36 @@
+HELP.md
```
아마 알고 계신데 깜박하신 것 같지만, `.gitignore` 파일 설정을 해두는 편이 나아 보입니다 🐱

epicblues replied:
혹시 어떤 파일을 추가하지 않아서 아쉬우셨는지 궁금합니다.

res-cogitans replied:
`.idea`의 `sonarlint 플러그인이나 `.gitingore`, `target` 등이 있을 것 같아요. `target`의 경우, 만약에 `restdocs` 관련 때문에 남겨두신 거라면 `generated-snippets`만 남길 수도 있을 것 같구요. (별개로, 이것도 자동적인 생성이 가능하기 때문에 굳이 남길 필요가 있나 싶긴 합니다.) 저 같은 경우에는 일일히 다 추가하기는 귀찮으니, 그냥 gitignore.io에서 자동 생성된 거 받은 다음에 수정하는 편이 편하더군요.

res-cogitans commented on src/main/java/com/programmers/epicblues/board/service/PostService.java:
```diff
@@ -0,0 +1,52 @@
+    this.userRepository = userRepository;
+  }
+
+  public List<PostResponse> getPosts(PageRequest pageRequest) {
```
Service의 모든 메서드들이 DTO인 `PostResponse`를 반환하고 있는데, 만약에 API 스펙이 변경되는 경우, 동일한 기능을 수행하되, 다른 DTO를 반환하는 메서드들이 그만큼 추가되게 되나요?

epicblues replied:
부실한 설계에 대해 변명을 하자면 생성, 수정, 조회 모두 클라이언트가 응답에 들어있는 Post에 대한 최신 정보를 바탕으로 작업을 할 것이라(SPA에서 화면에 렌더링 등) 생각이 들어서, 응답을 받고 또 다른 Post를 요청하는 일이 없도록 하기 위해 같은 DTO를 반환하도록 했습니다.

 API 스펙이 변경되거나 추가될 경우, 해당 요구사항에 맞는 DTO를 추가하거나, 급할 경우 그냥 Map으로 해결할 것 같습니다. 

 예를 들어 Post 삭제 API를 추가한다면 삭제된 Post의 id만 반환할 것 같습니다.

res-cogitans replied:
`Map` 사용이 간단하게 DTO 클래스 수를 줄여줄 수 있겠네요. 배워갑니다.

joseph-100 replied:
아뇨, 저는 Map 사용을 반대합니다. Map을 사용하면 어떤 데이터가 들어가 있는지 알기 어렵게 합니다. 데이터가 보이지 않으면 관리면에서 매우 불리해 집니다. 급하더라도 안됩니다.

res-cogitans commented on src/main/java/com/programmers/epicblues/board/service/PostService.java:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  public PostResponse createPost(Long userId, String title, String content) {
+    User user = userRepository.findById(userId).orElseThrow();
```
연관관계 설정을 위한 목적으로 `User` 엔티티를 조회하신 거라면, 프록시 조회 만으로도 같은 목적을 수행하면서도 DB 조회를 줄일 수 있습니다. (JPA책 8.1.2를 참고하시면 도움이 될 것 같습니다.) 데이터 JPA 기본 메서드 중에서 `em.getReference`와 같은 효과를 보시려면(즉 프록시 조회 하시려면) `getById` 메서드를 이용하시면 됩니다.

epicblues replied:
감사합니다! `getById`를 이렇게 사용하는 것이었군요!

joseph-100 replied:
@res-cogitans 여기서 프록시를 생각하신부분에 대해 저는 잘 이해가 되지 않네요. findById 로 조회해서 연관관계에 있는 post도 같이 조회 되진 않을 것 같아서요.

res-cogitans replied:
@joseph-100 User를 조회할 때, `findById` 사용해도 Post를 LAZY Loading 설정하였다면 당연히 프록시로 조회됩니다. 하지만 위의 코드의 경우 `Post`를 등록할 때 연관관계에 있는 `User`를 등록해주는 부분입니다.

1. `Post`에 `User`를 연관관계 맺어주려면, 영속 상태의 `User`가 필요하기에 `getById`건, `findById`건 간에 조회가 필요합니다.
2. 그런데, `Post` 영속화를 위한 연관관계 설정을 위해서는 `User`의 ID 값이면 충분하지(FK), `User` 엔티티의 필드들 값은 불필요합니다.
3. 때문에 위와 같은 연관관계 설정 상황이라면, `getById`로 프록시 객체를 불러와서 연관관계 설정에 사용하는 것이 SELECT 쿼리를 한 번 덜 날린다는 장점을 갖기 때문에 남긴 글이었습니다. (프록시의 경우, `getUserId()` 등 id 값 조회 시에는 굳이 초기화가 발생하지 않으니까요)

혹시 해서 다시 해 본 테스트 로그입니다:
`findById` 사용 시 다음과 같이 findById 부분에서 조회 쿼리가 나가는 것을 확인 가능합니다.
```
`2022-05-27 13:28:53.682  INFO 6304 --- [           main] o.p.s.service.post.PostServiceImpl       : userRepository에서 id값에 따라 User를 조회 시작`
Hibernate: 
    select
        user0_.user_id as user_id1_3_0_,
        user0_.created_date as created_2_3_0_,
        user0_.last_modified_date as last_mod3_3_0_,
        user0_.birth_date as birth_da4_3_0_,
        user0_.firstname as firstnam5_3_0_,
        user0_.lastname as lastname6_3_0_,
        user0_.nickname as nickname7_3_0_,
        user0_.hash_value as hash_val8_3_0_ 
    from
        user user0_ 
    where
        user0_.user_id=?
2022-05-27 13:28:53.718  INFO 6304 --- [           main] o.p.s.service.post.PostServiceImpl       : postRepository에 post를 저장 시작
```

반면, `getById` 사용 시 두 로그 사이에 조회 쿼리가 나가지 않는 것을 볼 수 있습니다.
```
2022-05-27 13:36:09.263  INFO 2632 --- [           main] o.p.s.service.post.PostServiceImpl       : userRepository에서 id값에 따라 User를 조회 시작
2022-05-27 13:36:09.373  INFO 2632 --- [           main] o.p.s.service.post.PostServiceImpl       : postRepository에 post를 저장 시작
```

res-cogitans commented on src/main/resources/application.yaml:
```diff
@@ -0,0 +1,17 @@
+      hibernate:
+        query.in_clause_parameter_padding: true
+  datasource:
+    url: jdbc:h2:~/test
```
h2 url설정을 메모리 모드로 변경해주시면, 외부 환경에서 다른 사람이 실행할 때도 H2가 설치되어 있지 않거나, 서버 파일을 새로 생성하지 않고도 바로 실행해볼 수 있습니다: 예를 들어 `jdbc:h2:mem:board` 하는 식으로 링크를 설정해 주시면 됩니다. (여기서 board는 무작위로 명명한 것입니다)

자세한 링크는 [여기](http://www.h2database.com/html/features.html)의 Database URL Overview 등을 참고하시면 됩니다.

epicblues replied:
오 기본값이 메모리 모드가 아니었군요. 배워갑니다✍️✍️

res-cogitans commented on src/main/java/com/programmers/epicblues/board/service/PostService.java:
```diff
@@ -0,0 +1,52 @@
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Transactional
```
이렇게 되면, 조회용 메서드들에도 `readOnly` 옵션은 없는 것이 되는데, 이 경우 모종의 실수로 오로지 조회용으로 만든 메서드 내부에서 실수로 변경이 발생했을 경우 DB로 반영이 되어, 의도치 않은 side-effect가 발생할 수 있기에 위험하다고 생각합니다. 

전역적으로는 `@Transactional(readOnly = true)`로 설정하되, 변경 반영이 필요한 메서드들에만 따로 readOnly를 false로 설정하는 것이 좋지 않을까 생각합니다.

epicblues replied:
스프링 시큐리티 강의에서 잭슨님도 조회 관련 메서드에는 `@Transactional(readOnly = true)`를 설정하셨던데 

작성할 때 주의해야 겠네요. 감사합니다!

res-cogitans commented on src/main/java/com/programmers/epicblues/board/service/PostService.java:
```diff
@@ -0,0 +1,52 @@
+    return PostResponse.from(savedPost);
+  }
+
+  public PostResponse getPostById(long postId) {
```
해당 메서드의 ID 값만 Wrapping 되지 않은 기본형 long을 사용하신 이유가 있으실까요?

epicblues replied:
으 실수입니다. ☠️☠️ 수정할게요

res-cogitans commented on src/main/java/com/programmers/epicblues/board/service/PostService.java:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  public PostResponse updatePost(Long postId, PostRequest postRequest) {
+    var targetPost = postRepository.findById(postId).orElseThrow();
```
`orElseThrow()`에 인자를 제공하지 않으면 `NoSuchElementException`이 발생하는 것은 처음 알았네요. 덕분에 배워갑니다 👍

res-cogitans commented on src/main/java/com/programmers/epicblues/board/controller/RestControllerExceptionHandler.java:
```diff
@@ -0,0 +1,38 @@
+public class RestControllerExceptionHandler {
+
+  @ResponseStatus(HttpStatus.BAD_REQUEST)
+  @ExceptionHandler(InvalidRequestArgumentException.class)
```
Controller Advice가 `InvalidRequestArgumentException`과 `ResourceNotFoundException`에 대해서만 Exception Handling 하고 있는데 혹시 그 외의 예외가 발생한다면 어떻게 되나요? 만약에 예상치 못한 `IllegalArgumentException`이 발생하는 경우 500 에러로 나가게 되나요? 가능하면 오류시 응답 스펙을 서버 쪽에서 맞춰 놓고 어떤 예외가 발생하더라도 응답하는게 낫지 않을까 하는 것이 개인적인 생각입니다. (예를 들어, `Exception`을 잡는 Exception Handler가 존재한다면 예상치 못한 오류가 나왔을 때도 정해진 메시지 포멧으로 메시지 전달이 가능하며, 혹은 규정한 오류 코드를 내보낼 수 있겠죠)

epicblues replied:
말씀하신대로 애플리케이션이 기본값으로 생성하는 에러 메시지에 의존하는 것보다는 직접 스펙을 정해놓는 것이 클라이언트의 혼란을 예방할 수 있겠네요. 수정하겠습니다

res-cogitans commented on src/main/java/com/programmers/epicblues/board/dto/PostPagePayload.java:
```diff
@@ -0,0 +1,24 @@
+@Getter
+public class PostPagePayload {
+
+  @NotNull(message = "page를 입력해야 합니다.")
```
message를 직접 입력해주실 수도 있지만 스프링의 메시지 기능을 이용하여 하나의 `properties` 파일 안에 에러 메시지를 일관되게 관리하실 수도 있습니다.

epicblues replied:
이런 용도로 사용하는 것이었군요. 따로 학습을 해봐야겠습니다. ✍️

res-cogitans commented on src/main/java/com/programmers/epicblues/board/entity/BaseEntity.java:
```diff
@@ -0,0 +1,20 @@
+@EntityListeners(AuditingEntityListener.class)
+public abstract class BaseEntity {
+
+  protected String createdBy;
```
`public`한 `Getter`를 제공하는데, 굳이 `protected`로 필드에 직접 접근할 수 있는 길을 열어 줄 필요는 없지 않을까 싶습니다.

epicblues replied:
저 같은 경우는 createdBy에 setter를  만들지 않아서

상속받은 User와 Post에서 createdBy를 수정하려면 protected 접근 제어자를 사용할 수 밖에 없어서 사용했는데

```java
 public void assignUser(User user) {
    this.user = user;
    this.createdBy = user.getName();
    if (!user.contains(this)) {
      user.addPost(this);
    }
  }
```

제가 어떤 개념 같은 것을 놓친 게 있을까요? setter를 꼭 설정해야 하나요?

res-cogitans replied:
아, set하는 거는 생각을 못 했네요. 최소한 createdBy는 설정이 필요하겠군요. 전 createdAt만 생각해서 댓글을 달았습니다.

res-cogitans commented on src/main/java/com/programmers/epicblues/board/entity/BaseEntity.java:
```diff
@@ -0,0 +1,20 @@
+  protected String createdBy;
+  @CreatedDate
+  @Column(updatable = false)
+  protected LocalDateTime createdAt;
```
최종 수정 일시도 추가하는 편이 문제 상황 발생 시 추적에 용이하다고 생각합니다. `@LastModifiedDate`

epicblues replied:
프로젝트 때는 꼭 추가해야겠군요!

res-cogitans commented on src/main/java/com/programmers/epicblues/board/entity/LongIdHolder.java:
```diff
@@ -0,0 +1,6 @@
+package com.programmers.epicblues.board.entity;
+
+public interface LongIdHolder {
```
이 부분은 취향 문제 같습니다만, `Getter`를 엔티티에서 같이 사용하고 계시기 때문에 만약에 `@Id`로 사용하는 필드명이 `id`에서 다른 이름으로 바뀔 경우 관리가 더 귀찮아지지 않을까요? 개인적으로는 이 인터페이스가 코드 수를 줄여주는 양적인 이점이 그리 커 보이지는 않아서요..

res-cogitans replied:
요건 테스트 코드 읽으면서 완전히는 모르겠지만, 대략 필요성은 배웠습니다 👍

epicblues replied:
테스트 전용 유틸 클래스인 FieldSetter를 위해 만들어둔 인터페이스인데, 말씀하신 것처럼 양적인 이점도 없고, 프로덕션 코드에 혼란만 만드는 것 같아서 삭제해야 겠습니다.

res-cogitans commented on src/main/java/com/programmers/epicblues/board/entity/Post.java:
```diff
@@ -0,0 +1,77 @@
+@Entity
+@Table(name = "post")
+@Getter
+@NoArgsConstructor
```
JPA 사용 시 `protected`이거나 `public`인 기본 생성자가 있어야 하는 것은 맞지만, 굳이 `public`하게 만들기 보다는 `protected`로 설정하는 것이 더 안전하다고 생각합니다: `@NoArgsConstructor(access = AccessLevel.PROTECTED)`

res-cogitans commented on src/main/java/com/programmers/epicblues/board/entity/User.java:
```diff
@@ -0,0 +1,59 @@
+  private int age;
+  private String hobby;
+  @OneToMany(mappedBy = "user", cascade = {CascadeType.PERSIST}, orphanRemoval = true)
+  private List<Post> posts = new ArrayList<>();
```
`List`와 같은 가변 객체 필드에 대해서 `public`한 `Getter`를 제공하는 것은 `Setter`를 `public`으로 열어 준 것과 같은 효과를 줍니다. `getPosts()` 메서드로 `List`에 대한 객체 참조를 얻은 것 만으로 `posts.add`, `posts.remove` 메서드를 호출할 수 있게 되니까요. (자세한 내용은 이펙티브 자바 아이템 17 참고하시면 도움 될 듯 합니다. 일급 컬렉션 사용도 고려해보시면 좋을 듯 합니다: 슬랙 [링크](https://greppkdt-backend.slack.com/archives/C026WAH3T7C/p1652840734620959))

epicblues replied:
List<Post>에 관한 public한 연산에 대해서 나름 신중했다(!) 생각했는데 getter를 방치하면 무용지물이 될 수도 있는 거였네요. 조언 감사합니다.
일단 `getPosts()`에서 `Collections.unmodifiableList(list)`를 반환하도록 수정했습니다.

res-cogitans commented on src/main/java/com/programmers/epicblues/board/entity/Post.java:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
JPA 프록시 사용하시는 경우 `equals`와 `hashCode` 구현할 때 필드 직접 접근이 아닌 Getter 이용하여 접근하게 만들어야 제대로 구현됩니다. (InteliJ에서 `equals`/`hashCode` 생성 기능을 사용할 때도 해당 옵션이 있습니다)

관련해서 제가 에전에 질문했던 글인데, 답변 내용을 읽어보시면 도움이 되실 듯 합니다: [링크](https://www.inflearn.com/questions/333337)

epicblues replied:
수정했습니다. 엔티티 클래스를 만들 때 신경 쓸 것이 많네요 🤣

res-cogitans commented on src/main/java/com/programmers/epicblues/board/entity/Post.java:
```diff
@@ -0,0 +1,77 @@
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(id, title, content);
```
`Object.hash`의 경우 `hashCode` 규약에는 맞습니다만 성능 상으로는 아쉬운 점이 있습니다. 관련해서 이펙티브 자바의 아이템 11, 특히 예제인 `코드11-3 한 줄짜리 hashCode 메서드 - 성능이 살짝 아쉽다` 주변의 내용을 살펴 보시면 좋을 듯 합니다.

epicblues replied:
한빈님의 코드를 봤을 때는 entity 클래스에 equals()와 hashCode()를 따로 재정의하지 않으셨던데, 그렇다면 꼭 필요한 상황이 아니면 차라리 재정의하지 않는 게 나을까요?

res-cogitans commented on src/main/java/com/programmers/epicblues/board/entity/Post.java:
```diff
@@ -0,0 +1,77 @@
+    return Objects.hash(id, title, content);
+  }
+
+  public void removeUser() {
```
이 부분은 생각이 갈릴 수 있을 것 같습니다만, 양방향 매핑을 굳이 사용하지 않아도 되는 곳에서는 사용하지 않는 편이 낫지 않을까 싶습니다. 

위의 연관관계 편의 메서드가 `User` 쪽이건, `Post`쪽이건 어느 쪽을 호출하건 간에 정상적으로 작동하기 위해서 이렇게 설계된 것 같은데, 가독성을 좀 떨어트리는 것 같아서요. 어느 쪽이건 호출해도 문제가 없게 만드느라 순환(에러는 안 일으키지만)이 발생하다보니 그렇게 된다고 느꼈습니다.

epicblues replied:
넵 프로젝트의 규모가 커져서 많은 엔티티를 다룰 때는 실수할 가능성이 클 것 같네요.

res-cogitans commented on src/main/java/com/programmers/epicblues/board/controller/PostController.java:
```diff
@@ -0,0 +1,89 @@
+      var postResponse = postService.getPostById(postId);
+      return ResponseEntity.ok(postResponse);
+
+    } catch (NoSuchElementException exception) {
```
개인적으로 여기서 예외 번역을 하지 않고, `orElseThrow`에서 던져줄 수도 있고, `NoSuchElementException`을 사용하고, 그걸 Exception Handling 한다거나, 방법이 다양하다고 생각합니다. 여기서 `ResourceNotFoundException`을 사용하신 까닭이 궁금합니다.

epicblues replied:
저는 제 애플리케이션에서만 발생할 수 있는 커스텀 예외만 ControllerAdvice에서 처리하도록 구현하고 싶어서 그렇게 했는데,  차라리 서비스에서 예외 번역을 했어야 하는 걸까요?

res-cogitans replied:
그런 거라면 확실히 고민이 되네요. 이 부분은 취향 같긴 한데, 저라면 서비스를 여러 컨트롤러(혹은 여러 버전)에서 사용해도 예외 종류에 일관성을 두기 위해 서비스 단에서 예외 번역을 하거나, 그냥 번역을 안 하거나 할 것 같습니다.

res-cogitans commented on src/test/java/com/programmers/epicblues/board/service/PostServiceTest.java:
```diff
@@ -0,0 +1,181 @@
+
+    // given
+    User user = getUser();
+    FieldSetter.assignId(user, 1L);
```
왜 민성님이 PR 코멘트에 ID 얘기를 작성하시고, IdHolder 인터페이스를 만드셨는지 좀 이해가 가는 부분이었습니다. JPA + Generated Id 사용하면 확실히 `Service`의 단위 테스트가 복잡해질수밖에 없네요. 공부가 되었습니다 :godmode: 

개인적으로는 `SpringBootTest`를 그냥 사용해서 통합 테스트를 하는 편이 실제 업무에서는 더 효율적인 방법이 되지 않을까 싶습니다.

res-cogitans commented on src/test/java/com/programmers/epicblues/board/entity/EntityJpaTest.java:
```diff
@@ -0,0 +1,130 @@
+import org.springframework.boot.test.context.SpringBootTest;
+import util.EntityFixture;
+
+/*
```
만약 리뷰 목적이 아닌 경우 프로젝트 디렉토리 안에 있는 학습 테스트에 가까운 코드들은 `.gitignore` 설정을 통해 공유하지 않는 것도 유용한 방법이라고 생각합니다.

epicblues replied:
그랬어야 했네요. 확실히 리뷰하는 사람 입장에서는 피곤할 것 같습니다.

res-cogitans replied:
개인적으로는 딱히 피곤하지는 않고 있어서 공부가 되기는 했는데, 저 같은 경우에는 배포하면 뭔가 찜찜해서(?) 일단 ingore처리를 했었거든요. 사실 저희 같은 학습 케이스면 뭐 크게 신경 쓰지는 않으셔도 될 듯 합니다. 나중에 숨기고 싶으실 때 생각해보심 될듯

res-cogitans commented on src/test/java/com/programmers/epicblues/board/controller/PostControllerTest.java:
```diff
@@ -0,0 +1,328 @@
+        content().json(json.writeValueAsString(expectedResponse))
+    ).andDo(document("post-get-page",
+        responseFields(
+            fieldWithPath("[].title").type(JsonFieldType.STRING).description("제목"),
```
이런 중첩되는 친구들은 `ignore`메서드나 `andWithPrefix` 메서드를 이용하여 귀찮게 치는 일을 줄일 수가 있습니다. 이 외에도 아름답게(?) 재사용할 수 있는 많은 방법을 RESTDocs에서 제공하고 있던데, 저도 완벽한 이해가 된 상태가 아닌지라 링크로 갈음합니다: [링크](https://docs.spring.io/spring-restdocs/docs/current/reference/html5/)

일단 위의 공식링크에 이런 귀찮은 코드 작성을 방지하기 위한 방법들이 많이 소개되어 있습니다. 나중에 시간 나시거나 필요하실 때 보시면 좋을듯 합니다  📚

res-cogitans reviewed (commented):
민성님, 코드 잘 봤습니다! 👍 

코드 읽으면서 들었던 제 의견들, 궁금했던 부분들 코멘트로 남겼습니다.

테스트 부분에 워낙 새로운 것들이 많다 보니 다 이해는 못해서 리뷰를 그 쪽으로는 그리 못 남겨서 죄송스럽네요 .. 오히려 제가 더 공부가 되었습니다 :godmode: 

혹여 리뷰에 이상한 점이라던가, 의문점 등 있으시면 편하게 말씀 주세요! 👋

epicblues committed:
fix : 게시글 생성 연관관계 설정 오버헤드 문제

JpaRepository.findById() -> JpaRepository.getById();

epicblues committed:
fix : 리뷰 반영

JpaRepository.findById() -> JpaRepository.getById();

최종 예외 처리 함수

LongIdHolder 인터페이스 삭제

Transactional readonly 설정

h2 데이터베이스 in-memory로 작동하도록 변경

joseph-100 commented on src/main/java/com/programmers/epicblues/board/service/PostService.java:
```diff
@@ -0,0 +1,55 @@
+
+  @Transactional
+  public PostResponse createPost(Long userId, String title, String content) {
+    User user = userRepository.getById(userId);
```
어떤것은 var를 쓰고 어떤건 클래스 타입을 씁니다. 통일해서 사용하시면 좋을거 같아요.

epicblues replied:
클래스 타입으로 통일했습니다!

joseph-100 commented on src/main/java/com/programmers/epicblues/board/service/PostService.java:
```diff
@@ -0,0 +1,55 @@
+  public PostResponse updatePost(Long postId, PostRequest postRequest) {
+    var targetPost = postRepository.findById(postId).orElseThrow();
+    targetPost.updateTitle(postRequest.getTitle());
+    targetPost.updateContent(postRequest.getContent());
```
그냥 update(postRequest); 로 한번에 해도 될 거 같아요.

epicblues replied:
수정했습니다 !UpdatePostRequest라는 새로운 DTO를 만들었습니다!

joseph-100 commented on src/main/java/com/programmers/epicblues/board/controller/PostController.java:
```diff
@@ -0,0 +1,89 @@
+      throw new InvalidRequestArgumentException(bindingResult);
+    }
+
+    var createdResult = postService.createPost(postRequest.getUserId(), postRequest.getTitle(),
```
createPost 에 그냥 postRequest를 넘기면 안되나요?

epicblues replied:
수정했습니다! CreatePostRequest라는 새로운 DTO를 만들었습니다!

joseph-100 commented on src/main/java/com/programmers/epicblues/board/controller/PostController.java:
```diff
@@ -0,0 +1,89 @@
+
+  @GetMapping("/posts")
+  public ResponseEntity<List<PostResponse>> getPostWithPage(
+      @Valid @ModelAttribute PostPagePayload payload, BindingResult bindingResult) {
```
컨트롤러 파라미터로 Pageable 객체를 받을 수 있습니다. 사용법을 한번 찾아보세요.

epicblues replied:
수정했습니다! 다만 parameter로 page나 size가 잘못 입력 됐을 경우 사용자에게 400 Bad Request 응답을 보내고 싶은데 , 따로 validation을 하지 않고 기본 값으로 Pageable 구현체를 매개변수 넘겨주는 것 같아서 아쉽네요!😥

joseph-100 commented on src/main/java/com/programmers/epicblues/board/controller/PostController.java:
```diff
@@ -0,0 +1,89 @@
+  private final PostService postService;
+
+  public PostController(PostService postService) {
+    this.postService = postService;
```
Lombok 을 사용하시면 @RequiredArgsConstructor 로 대체하실수 있습니다.

epicblues replied:
수정했습니다! final 로 선언된 필드 뿐만 아니라 @NonNull과 같은 constraints가 달린 필드들도 포함시키는 것이 신기하네요!

joseph-100 reviewed (commented):
민성님 과제하시느라 고생하셨습니다.
코드 작성에 더 신경쓰는 느낌이 들어 좋습니다. 계속 더 신경 써 주세요.
테스트 커버리지 100%는 놀랍네요 ㅎㅎ

1. 멘토님은 메서드에서 예외가 발생하는 상황을 테스트 하실 때 미리 예외 클래스를 설계하시는지 궁금합니다!
- 반드시 필요할 예측 가능한 선에서는 미리 예외를 만들어 두긴 하지만 보통은 필요에 따라서 만드는 편입니다.
2. 예외 클래스를 작성하실 때 애플리케이션에서 최상위 예외 클래스를 먼저 작성하시는 지도 궁금합니다.
저는 생각이 잘 나지 않아서 일단 모두 RuntimeException을 상속 받아서 작성했는데, 먼저 어느 정도 구조를 짜고 들어갔어야 했을까요?
- 질문을 잘 이해하지 못했습니다. 애플리케이션 내 커스텀하게 예외를 만들 경우는 대부분 Runtime 예외 겠지요. 커스텀한 예외를 따로 다루기 위해서 RuntimeException 을 상속한 기본 예외를 만들고, 그 하위로 예외를 만들어서 사용하기도 합니다.
3. TODO 1 관련 : 시간을 직렬화할 때 format을 프로젝트 별로 따로 작성하시는지, 아니면 업계의 표준(?) 같이 자주 쓰는 형식이 있는지 궁금합니다!
- 니즈에 따라 다를거 같네요. 프로젝트내 에서 시간에 대해 동일한 포맷으로 맞출 수 있으면 공통설정으로 두어도 되지만, 저는 외부 시스템과 연동이 많아 필드별로 각각 포맷을 정의합니다.
- 표준이란건 없지만 제 경험상으론 민성님이 설정하신것에 24시간 표기(HH)로 바꿔 사용을 많이해요.
4. TODO 2 관련 : 테스트 전용 util 클래스에 대한 기능을 테스트하고 싶어서 해당 클래스의 내부 클래스에 테스트를 작성했는데, 멘토님은 따로 테스트 클래스 파일을 작성하시는지 궁금합니다!
- 열정은 좋습니다. 저는 자바 리플렉션 코드에 대한 의심을 갖지 않아서 해당 코드에 대해서는 별도로 만들것 같진 않아요.
5. TODO 3 관련 : 저번 바우처 과제에서 Controller를 테스트 할 때 Service를 모의 객체를 만들어서 테스트 코드를 작성했으나, 이번 과제에서는 실제 작동하는 service를 이용해서 통합 테스트로 작성했습니다. 이유는 이렇습니다.
- 어제 팀 회의때 답변을 드린것 같네요. 저는 서비스를 목처리 할 것 같아요. 생각하신 내용 잘 정리해서 남겨주신 부분은 아주 좋습니다.

epicblues committed:
refactor : 코드 리뷰 반영 2

BaseEntity
* @PrePersist로 수동 auditing

CreatePostRequest
* 게시글 생성 요청 전담 DTO

UpdatePostRequest
* 게시글 수정 요청 전담 DTO

PostPagePayload.java 삭제
* 스프링의 paging 요청 관련 ArgumentResolver로 대체

PostService
* 메서드 시그니처 매개변수 Value Object들로 변경

epicblues committed:
fix : POST /posts

PostController, PostControllerTest
* 만들어진 컨텐츠를 Response Body에 넣어서 보내는 대신
생성된 자원의 URI를 Location 헤더에 할당
* 응답 상태 코드 201로 변경
